### ISSUE TITLE
Question about browser versions

### ISSUE THREAD

**Issue by ShoopIndy** [zalando/zalenium#244]
Apologies if there is a more appropriate place for usage questions but I couldn't find it.

However, I’m not sure if I understand the appropriate way to run tests using an older browser version (without using an external service).

As far as I can tell, the only way to do this is to set the --seleniumImageName option (when starting zalenium) to the appropriate tag of elgalu/selenium that contains the desired version of chrome.  For example, if I want to use version 59.0.3071.115 of chrome, then I would pull and set seleniumImageName to ‘elgalu/selenium:3.3.1-p18’.  I believe that the elgalu/selenium images used to contain multiple browser versions (which could be used by setting an environment variable) but I don’t believe that this is the case anymore.

Is this correct or is there a simpler way to run zalenium (elgalu/selenium) with an older browser version? (outside of using an external service like Browserstack).

**Comment by elgalu** [comment]
All your statements are correct @ShoopIndy 

Following your example, once you choose `elgalu/selenium:3.3.1-p18` you should also choose a Zalenium version `3.3.1` so Selenium versions are compatible with each other.

What I would do instead is to build my own `Dockerfile` and overwrite the Chrome version I need:

```
FROM elgalu/selenium
# TODO: Here I overwrite the installed chrome with the older one
```

This is the reason why I included the Chrome deb files in the releases, so a user like you could download that .deb file and install the exact version the need. Maybe there is a better way but I couldn't find it.

So those seem to be the current options.

**Comment by ShoopIndy** [comment]
@elgalu Thanks for the help.

Installing the desired browser is definitely a better way to do this since I would still get any improvements in the latest builds.  However, I’m running into problems with doing this.

I created a dockerfile containing nothing more than:
```
FROM elgalu/selenium:3.3.1-p27
ENV FFMPEG_FRAME_RATE 2
```
After confirming that this worked, I added commands to un-install and re-install the same version of chrome that was originally in the image (61.0.3163.91) by basically just re-using the commands from the base image [Dockerfile](https://github.com/elgalu/docker-selenium/blob/3.3.1-p27/Dockerfile#L583-L611):

```
FROM elgalu/selenium:3.3.1-p27

ENV FFMPEG_FRAME_RATE 2

ARG CHROME_VER="61.0.3163.91"

#ENV CHROME_URL="https://sourceforge.net/projects/chromehistory/files"  # versions 43-51
ENV CHROME_URL="https://dl.google.com/linux/direct"

ENV CHROME_VERSION_TRIGGER="$CHROME_VER"
ENV CHROME_BASE_DEB_PATH="/home/seluser/chrome-deb/"
ENV GREP_ONLY_NUMS_VER="[0-9.]{2,20}"

LABEL selenium_chrome_version "$CHROME_VER"
LABEL selenium3_chrome_version "$CHROME_VER"

#ENV CHROME_DEB="google-chrome-stable_${CHROME_VER}_amd64.deb"  # older versions
ENV CHROME_DEB_NAME="google-chrome-stable_current_amd64.deb"

ENV CHROME_DEB_PATH="${CHROME_BASE_DEB_PATH}${CHROME_DEB_NAME}"

USER root

RUN apt-get -y remove google-chrome-stable
RUN rm /opt/google/chrome/google-chrome-base

RUN apt-get -qqy update
RUN mkdir -p chrome-deb
RUN wget -nv "${CHROME_URL}/${CHROME_DEB_NAME}" -O "./chrome-deb/${CHROME_DEB_NAME}"
RUN apt-get -qyy --no-install-recommends install "${CHROME_DEB_PATH}"
RUN rm "${CHROME_DEB_PATH}"
RUN rm -rf ./chrome-deb
RUN apt-get -qyy autoremove
RUN rm -rf /var/lib/apt/lists/*
RUN apt-get -qyy clean
RUN export CH_STABLE_VER=$(/usr/bin/google-chrome-stable --version | grep -iEo "${GREP_ONLY_NUMS_VER}")
RUN echo "${CH_STABLE_VER}"

# We have a wrapper for /opt/google/chrome/google-chrome
RUN mv /opt/google/chrome/google-chrome /opt/google/chrome/google-chrome-base
```
However, after I rebuild this image and try using it with Zalenium, I get selenium Net::ReadTimeout errors and the following zalenium logs:

```
Docker binary already present, will use that one.
Docker version 17.03.0-ce, build 60ccb22
-- LOG 21:23:47:998108500 Ensuring docker works...
-- LOG 21:23:48:047203423 Ensuring no stale Zalenium related containers are still around...
-- LOG 21:23:48:071141753 Ensuring docker-selenium is available...
-- LOG 21:23:48:294680572 Running additional checks...
Copying files for Dashboard...
cp: cannot create regular file '/home/seluser/videos/index.html': No such file or directory
Starting Nginx reverse proxy...
Starting Selenium Hub...
21:23:49.098 INFO - Selenium build info: version: '3.3.1', revision: '5234b32'
21:23:49.099 INFO - Launching Selenium Grid hub
21:23:49.937 INFO - binding de.zalando.ep.zalenium.servlet.LivePreviewServlet to /grid/admin/LivePreviewServlet/*
21:23:49.940 INFO - binding de.zalando.ep.zalenium.servlet.ZaleniumConsoleServlet to /grid/admin/ZaleniumConsoleServlet/*
21:23:49.943 INFO - binding de.zalando.ep.zalenium.servlet.ZaleniumResourceServlet to /grid/admin/ZaleniumResourceServlet/*
21:23:49.945 INFO - binding de.zalando.ep.zalenium.dashboard.DashboardCleanupServlet to /grid/admin/DashboardCleanupServlet/*
21:23:50.142 INFO - Will listen on 4445
21:23:50.462 INFO - Nodes should register to http://172.18.0.2:4445/grid/register/
21:23:50.462 INFO - Selenium Grid hub is up and running
........Selenium Hub started!
Starting DockerSeleniumStarter node...
21:23:51.295 INFO - Selenium build info: version: '3.3.1', revision: '5234b32'
21:23:51.297 INFO - Launching a Selenium Grid node
2017-09-19 21:23:52.128:INFO::main: Logging initialized @1352ms to org.seleniumhq.jetty9.util.log.StdErrLog
21:23:52.201 INFO - Driver provider org.openqa.selenium.ie.InternetExplorerDriver registration is skipped:
 registration capabilities Capabilities [{ensureCleanSession=true, browserName=internet explorer, version=, platform=WINDOWS}] does not match the current platform LINUX
21:23:52.202 INFO - Driver provider org.openqa.selenium.edge.EdgeDriver registration is skipped:
 registration capabilities Capabilities [{browserName=MicrosoftEdge, version=, platform=WINDOWS}] does not match the current platform LINUX
21:23:52.203 INFO - Driver class not found: com.opera.core.systems.OperaDriver
21:23:52.203 INFO - Driver provider com.opera.core.systems.OperaDriver registration is skipped:
Unable to create new instances on this machine.
21:23:52.204 INFO - Driver class not found: com.opera.core.systems.OperaDriver
21:23:52.204 INFO - Driver provider com.opera.core.systems.OperaDriver is not registered
21:23:52.206 INFO - Driver provider org.openqa.selenium.safari.SafariDriver registration is skipped:
 registration capabilities Capabilities [{browserName=safari, version=, platform=MAC}] does not match the current platform LINUX
2017-09-19 21:23:52.289:INFO:osjs.Server:main: jetty-9.2.20.v20161216
2017-09-19 21:23:52.386:INFO:osjsh.ContextHandler:main: Started o.s.j.s.ServletContextHandler@4b5d6a01{/,null,AVAILABLE}
2017-09-19 21:23:52.417:INFO:osjs.AbstractConnector:main: Started ServerConnector@32479910{HTTP/1.1,[http/1.1]}{0.0.0.0:30000}
2017-09-19 21:23:52.418:INFO:osjs.Server:main: Started @1642ms
21:23:52.418 INFO - Selenium Grid node is up and ready to register to the hub
21:23:52.442 INFO - Registering the node to the hub: http://localhost:4444/grid/register
21:23:52.896 INFO - SessionCleaner initialized with insideBrowserTimeout 0 and clientGoneTimeout 1800000 polling every 180000
21:23:55.128 INFO - [DS] Capabilities grabbed from the docker-selenium image
21:23:55.147 INFO - The node is registered to the hub and ready to use
21:23:55.158 INFO - Registered a node http://172.18.0.2:30000
21:23:55.159 INFO - [DS]  Setting up 2 nodes...
21:23:55.207 INFO - 2 containers were created, it will take a bit more until all get registered.
................DockerSeleniumStarter node started!
Sauce Labs not enabled...
Browser Stack not enabled...
TestingBot not enabled...
Zalenium is now ready!
*************************************** Data Processing Agreement ***************************************
By using this software you agree that the following non-PII (non personally identifiable information)
data will be collected, processed and used by Zalando SE for the purpose of improving our test
infrastructure tools. Anonymisation with respect of the IP address means that only the first two octets
of the IP address are collected.

See the complete license at https://github.com/zalando/zalenium/blob/master/LICENSE.md
*************************************** Data Processing Agreement ***************************************
21:24:13.096 INFO - Registered a node http://172.18.0.3:40001
21:24:14.139 INFO - [DS] Container zalenium_40001 is up after ~14 seconds...
21:24:15.967 INFO - Got a request to create a new session: Capabilities [{rotatable=false, nativeEvents=false, name=Browser_Information-Capture_browser_information, browserName=chrome, takesScreenshot=false, javascriptEnabled=false, screenResolution=1280x720, version=, platform=ANY, cssSelectorsEnabled=false, group=Browser_Information}]
21:24:15.973 INFO - [DS] Starting new node for {rotatable=false, nativeEvents=false, tz=Europe/Berlin, name=Browser_Information-Capture_browser_information, browserName=chrome, takesScreenshot=false, javascriptEnabled=false, screenResolution=1280x720, version=, platform=ANY, cssSelectorsEnabled=false, group=Browser_Information}.
21:24:59.300 INFO - [DS] Container zalenium_40000 took longer than 60 seconds to be up and ready, this might be a signal that you have reached the hardware limits for the number of concurrent threads that you want to execute.
```
Am I not installing chrome properly?

p.s. I’m using dosel/zalenium:3.3.1q, egalu/selenium:3.3.1-p27 and selenium-webdriver 3.3.0.

**Comment by elgalu** [comment]
Note `CHROME_DEB_NAME="google-chrome-stable_current_amd64.deb"` will always install latest chrome.

The labels

```
LABEL selenium_chrome_version "$CHROME_VER"
LABEL selenium3_chrome_version "$CHROME_VER"
```

Are just for internal info. So in that script `CHROME_VER` has no impact regarding chaning Chrome version.

Independently of that I will try it myself to see why docker-selenium failed to start.

**Comment by elgalu** [comment]
@ShoopIndy I managed to install an older browser and pushed that Dockerfile as an example, please check https://github.com/elgalu/docker-selenium/blob/master/Dockerfile.older.browser

**Comment by akvsh** [comment]
@elgalu I tried what was recommended above for using older browser versions and currently my hub/pods fail to spin up. I wanted to use firefox 47.0.1 so I used `elgalu/selenium:2.53.1` and `dosel/zalenium:2.53.1u` in my Deployment yaml file (I am using the kubernetes version). Originally when I was using the latest version of zalenium the pods would atleast spin up, but the tests would still not run.

The only output in the hub pod's logs currently is 
```
Docker binary already present, will use that one.
Docker version 17.03.0-ce, build 60ccb22
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```
